ssl/tls13_server.cc

```diff
@@ -826,7 +826,14 @@ static enum ssl_hs_wait_t do_send_server_hello(SSL_HANDSHAKE *hs) {
 
   if (!ssl->s3->session_reused) {
     // Determine whether to request a client certificate.
-    hs->cert_request = !!(hs->config->verify_mode & SSL_VERIFY_PEER);
+    hs->cert_request = !!(hs->config->verify_mode & SSL_VERIFY_PEER) &&
+                       !(hs->config->verify_mode & SSL_VERIFY_POST_HANDSHAKE);
+
+    // Client authentication requested but immediately after initial handshake
+    if(!!(hs->config->verify_mode & SSL_VERIFY_POST_HANDSHAKE) &&
+        !!(hs->config->verify_mode & SSL_VERIFY_PEER)) {
+      ssl->s3->pha_ext = SSL_PHA_REQUEST_PENDING;
```

Comment:
* I'm not quite sure if the original logic in `!!(hs->config->verify_mode & SSL_VERIFY_PEER)` was intentional, or if we could just use `(hs->config->verify_mode & SSL_VERIFY_PEER)` 
* Could you help clarify what the behavior is supposed to be in this if statement? I blame the double exclamation marks, but I'm a bit confused :(  Is this supposed to be configured if both `SSL_VERIFY_POST_HANDSHAKE` and `SSL_VERIFY_PEER` are on?

Reply:
- The double exclamation converts a given value into the same truth value in a boolean format (ex. taking an int and then applying the !! operator will give you a boolean representing the same truth value). In this context, the & operator returns the binary addition of the two values provided (which could be any int, 10 & 12 would return 8). I assume the !! is just to explicitly ensure we are converting the result into a boolean value.
- The behavior before was if SSL_VERIFY_PEER is set, then hs->cert_request should be true. This makes sense since if the server sets this flag, they want to send a certificate request in the handshake. With the introduction of SSL_VERIFY_POST_HANDSHAKE, we allow servers to modify this behavior. If both flags are set together, then the server will authenticate the client but right after the initial handshake. So here, only when both flags are set, we set pha_ext to SSL_PHA_REQUEST_PENDING and leave hs->cert_request to false such that the server does not send a certificate request in the initial handshake but does so right after it.

Reply:
I'll add in some of this explanation explicitly in the comments. As for the exclamation marks I agree, they make the logic confusing. I wasn't sure if they were put in for a reason, but I can simplify it since all we care about is the value not whether it's truly boolean or not.